### PR TITLE
[search][generator] External postcodes.

### PR DESCRIPTION
**Pull request by vng**
Postcodes for US and UK (GB).

Fixes https://github.com/organicmaps/organicmaps/issues/717
A partial fix for https://github.com/organicmaps/organicmaps/issues/1876

### CODE REVIEW

File: generator/postcode_points_builder.cpp [hunk 1]
```diff
@@ -84,77 +86,53 @@ void GetPostcodes(std::string const & filename, storage::CountryId const & count
     if (std::find(countries.begin(), countries.end(), countryId) == countries.end())
       continue;
 
-    auto postcode = fields[fieldIndices.m_postcodeIndex];
-    if (!transformPostcode(postcode))
-      continue;
-
-    CHECK_EQUAL(valueMapping.size(), index, ());
+    keyValuePairs.emplace_back(search::NormalizeAndSimplifyString(fields[fieldIndices.m_postcodeIndex]),
```
Comment (RedAuburn) [inline generator/postcode_points_builder.cpp:89]: Can you capitalize all the letters in postcodes? Lowercase postcodes look strange:


i can't find any postcode formats where lowercase letters are used: https://en.wikipedia.org/wiki/List_of_postal_codes


(thank you so much for working on this, i think it'll make a lot of people happy!)

File: base/string_utils.cpp [hunk 2]
```diff
@@ -232,6 +233,21 @@ void Trim(std::string_view & sv)
     sv = {};
 }
 
+void Trim(std::string_view & s, std::string_view anyOf)
```
Comment (biodranik) [inline base/string_utils.cpp:236]: nit: returning sv here and above may work slightly faster and have less braces.

File: base/string_utils.cpp [hunk 3]
```diff
@@ -232,6 +233,21 @@ void Trim(std::string_view & sv)
...
+{
+  auto i = s.find_first_not_of(anyOf);
+  if (i != std::string_view::npos)
+  {
+    s.remove_prefix(i);
+
+    i = s.find_last_not_of(anyOf);
+    ASSERT(i != std::string_view::npos, ());
```
Comment (biodranik) [inline base/string_utils.cpp:244]: nit: This assert looks unnecessary.

File: generator/search_index_builder.cpp [hunk 4]
```diff
@@ -648,9 +648,14 @@ bool BuildSearchIndexFromDataFile(std::string const & country, feature::Generate
       writer->Seek(endOffset);
     }
 
+    /// @todo FilesContainerW::Write with section overriding invalidates current container instance
+    /// (@see FilesContainerW::GetWriter), thus we can't make 2 consecutive Write calls.
```
Comment (biodranik) [inline generator/search_index_builder.cpp:652]: Should this be merged into the master separately ASAP?

File: tools/python/maps_generator/var/etc/map_generator.ini.default [hunk 5]
```diff
@@ -76,6 +76,8 @@ NEED_BUILD_WORLD_ROADS: false
 # FOOD_TRANSLATIONS_URL:
 # SRTM_PATH:
 # ISOLINES_PATH:
+
+# Local path (not url!) to .csv files.
 # UK_POSTCODES_URL:
```
Comment (biodranik) [inline tools/python/maps_generator/var/etc/map_generator.ini.default:81]:
```suggestion
# UK_POSTCODES_PATH:
```

**Review by biodranik — APPROVED** [review]
LGTM, let's betatest it.

File: generator/generator_tool/generator_tool.cpp [hunk 6]
```diff
@@ -387,23 +376,13 @@ MAIN_WITH_ERROR_HANDLING([](int argc, char ** argv)
...
-          LOG(LCRITICAL,
-              ("Countries file is needed. Please set countries file name (countries.txt). "
-               "File must be located in data directory."));
-          return EXIT_FAILURE;
-        }
-
-        auto const topmostCountry = (*countryParentGetter)(country);
         bool res = true;
-        if (topmostCountry == "United Kingdom" && !FLAGS_uk_postcodes_dataset.empty())
+        if (!FLAGS_uk_postcodes_dataset.empty() && strings::StartsWith(country, "UK_"))
```
Comment (biodranik) [inline generator/generator_tool/generator_tool.cpp:380]: Do you have any idea on how to pass any other external databases with postal codes in the same format without introducing a command-line flag for each country? E.g. if I want to add swiss postal codes? Maybe simply pass a folder with files starting with prefixes matching our countries?

File: generator/postcode_points_builder.cpp [hunk 7]
```diff
@@ -84,77 +86,53 @@ void GetPostcodes(std::string const & filename, storage::CountryId const & count
...
 }
 
-template <typename Key, typename Value>
 void GetUSPostcodes(std::string const & filename, storage::CountryId const & countryId,
                     storage::CountryInfoGetter const & infoGetter,
-                    std::vector<m2::PointD> & valueMapping, std::vector<std::pair<Key, Value>> & keyValuePairs)
+                    std::vector<m2::PointD> & valueMapping, std::vector<ZipIndexValue> & keyValuePairs)
 {
-  // Zip;City;State;Latitude;Longitude;Timezone;Daylight savings time flag;geopoint
+  // * Get data source from here: https://simplemaps.com/data/us-zips
```
Comment (biodranik) [inline generator/postcode_points_builder.cpp:116]: Is all attribution present in data/copyright.html? @rtsisyk can you please double-check it?

File: search/postcode_points.cpp [hunk 8]
```diff
@@ -16,10 +14,12 @@
 
 #include "defines.hpp"
 
-using namespace std;
 
 namespace search
 {
+using namespace std;
```
Comment (biodranik) [inline search/postcode_points.cpp:20]: meh...

**Review by biodranik — APPROVED** [review]

File: generator/postcode_points_builder.cpp (hunk 7 above)
Comment (vng) [reply generator/postcode_points_builder.cpp:116]: In theory we can use Nominatim and not simplemaps.com, but:
```
По поводу Nominatim dataset. Первый же entry по US оттуда: 00586,18.343697,-67.028434
Google Maps говорит что в этих координатах 00685
В simplemaps.com такого кода нет.
Google Maps такой код находит только в Африке.
```

**Comment by vng** [comment]
@RedAuburn 
Is "WC2" https://overpass-turbo.eu/s/1LQt a real postal code?
I don't see it on the nominatim data source.

**Comment by RedAuburn** [comment]
> @RedAuburn Is "WC2" https://overpass-turbo.eu/s/1LQt a real postal code? Or it is used like a _prefix_? I don't see it on the nominatim data source, only WC2A .. WC2R

It's a postcode district, which contains prefixes inside like WC2A etc. Doesn't need search support IMO 👍️ https://en.wikipedia.org/wiki/WC_postcode_area

**Pull request merged**